File: typescript/structual-patterns/composite.ts
abstract class Component {
  public abstract search(keyword: string): void;
}

class MyFile extends Component {
  protected name: String;

  constructor(name: String) {
    super();
    this.name = name;
  }

  public search(keyword: string): void {
    console.log(`Searching for keyword ${keyword} in file: ${this.name}`);
  }
}

class Folder extends Component {
  protected name: String;
  protected components: Component[] = [];

  constructor(name: String) {
    super();
    this.name = name;
  }

  public add(component: Component): void {
    this.components.push(component);
  }

  public search(keyword: string): void {
    console.log(`Searching for keyword ${keyword} in folder: ${this.name}`);
    for (const component of this.components) {
      component.search(keyword);
    }
  }
}

function clienCode(component: Component, keyword: string): void {
  component.search(keyword);
}

const file1 = new MyFile("File 1");

file1.search("test");

console.log("----------");

const folder1 = new Folder("Folder 1");
const folder2 = new Folder("Folder 2");
const file2 = new MyFile("File 2");
const file3 = new MyFile("File 3");

folder2.add(file3);

folder1.add(file1);
folder1.add(file2);
folder1.add(folder2);

folder1.search("test");

File: rust/src/bin/structual-patterns/composite.rs
trait Component {
    fn search(&self, keyword: &str);
}

struct File {
    name: String
}

impl File {
    fn new(name: String) -> Self {
        File { name }
    }
}

impl Component for File {
    fn search(&self, keyword: &str) {
        println!("Searching for keyword '{}' in file: '{}'", keyword, self.name);
    }
}

struct Folder {
    name: String,
    components: Vec<Box<dyn Component>>
}

impl Folder {
    fn new(name: String) -> Self {
        Folder { name, components: Vec::new() }
    }

    fn add_component(&mut self, component: Box<dyn Component>) {
        self.components.push(component);
    }
}

impl Component for Folder {
    fn search(&self, keyword: &str) {
        println!("Searching for keyword '{}' in folder: '{}'", keyword, self.name);

        for component in self.components.iter() {
            component.search(keyword);
        }
    }
}

fn call<T: Component>(component: T, keyword: &str) {
    component.search(keyword);
}

fn main() {
    let file1 = File::new("File 1".to_string());

    call(file1, "test");

    println!("-----------");

    let mut folder1 = Folder::new("Folder 1".to_string());
    let mut folder2 = Folder::new("Folder 2".to_string());
    let file2 = File::new("File 2".to_string());
    let file3 = File::new("File 3".to_string());
    let file4 = File::new("File 4".to_string());

    folder1.add_component(Box::new(file2));
    folder1.add_component(Box::new(file3));
    folder2.add_component(Box::new(file4));

    folder1.add_component(Box::new(folder2));

    call(folder1, "test");
}